x-pack/plugins/security_solution/public/onboarding/components/onboarding_body/cards/integrations/utils.ts

```diff
@@ -0,0 +1,128 @@
+    .then((pkg) => pkg.PackageListGrid),
+}));
+
+export const fetchAvailablePackagesHook = (): Promise<AvailablePackagesHookType> =>
+  import('@kbn/fleet-plugin/public')
+    .then((module) => module.AvailablePackagesHook())
+    .then((hook) => hook.useAvailablePackages);
+
+export const extractFeaturedCards = (
+  filteredCards: IntegrationCardItem[],
+  featuredCardNames?: string[]
+) => {
+  return filteredCards.reduce((acc: Record<string, IntegrationCardItem>, card) => {
+    if (featuredCardNames?.includes(card.name)) {
+      acc[card.name] = card;
+    }
+    return acc;
+  }, {});
+};
+
+export const getFilteredCards = ({
+  customCardNames,
+  getAppUrl,
+  installedIntegrationList,
+  integrationsList,
+  navigateTo,
+}: {
+  customCardNames?: string[];
+  getAppUrl: GetAppUrl;
+  installedIntegrationList?: IntegrationCardItem[];
+  integrationsList: IntegrationCardItem[];
+  navigateTo: NavigateTo;
+}) => {
+  const securityIntegrationsList = integrationsList.map((card) =>
+    addSecuritySpecificProps({ navigateTo, getAppUrl, card, installedIntegrationList })
+  );
+  if (!customCardNames) {
+    return { featuredCards: {}, integrationCards: securityIntegrationsList };
+  }
+
+  return {
+    featuredCards: extractFeaturedCards(securityIntegrationsList, customCardNames),
+    integrationCards: securityIntegrationsList,
+  };
+};
+
+const addPathParamToUrl = (url: string, onboardingLink: string) => {
+  const encoded = encodeURIComponent(onboardingLink);
+  const paramsString = `${ONBOARDING_LINK}=${encoded}&${ONBOARDING_APP_ID}=${APP_UI_ID}`;
+
+  if (url.indexOf('?') >= 0) {
+    return `${url}&${paramsString}`;
+  }
+  return `${url}?${paramsString}`;
+};
+
+const addSecuritySpecificProps = ({
+  navigateTo,
+  getAppUrl,
+  card,
+}: {
+  navigateTo: NavigateTo;
+  getAppUrl: GetAppUrl;
+  card: IntegrationCardItem;
+  installedIntegrationList?: IntegrationCardItem[];
+}): IntegrationCardItem => {
+  const onboardingLink = getAppUrl({ appId: SECURITY_UI_APP_ID, path: ONBOARDING_PATH });
+  const integrationRootUrl = getAppUrl({ appId: INTEGRATION_APP_ID });
+  const state = {
+    onCancelNavigateTo: [APP_UI_ID, { path: ONBOARDING_PATH }],
+    onCancelUrl: onboardingLink,
+    onSaveNavigateTo: [APP_UI_ID, { path: ONBOARDING_PATH }],
+  };
+  const url =
+    card.url.indexOf(APP_INTEGRATIONS_PATH) >= 0 && onboardingLink
+      ? addPathParamToUrl(card.url, onboardingLink)
+      : card.url;
+  return {
+    ...card,
+    titleLineClamp: CARD_TITLE_LINE_CLAMP,
+    descriptionLineClamp: CARD_DESCRIPTION_LINE_CLAMP,
+    maxCardHeight: MAX_CARD_HEIGHT,
+    showInstallationStatus: true,
+    url,
+    onCardClick: () => {
+      if (url.startsWith(APP_INTEGRATIONS_PATH)) {
+        navigateTo({
+          appId: INTEGRATION_APP_ID,
+          path: url.slice(integrationRootUrl.length),
+          state,
+        });
+      } else if (url.startsWith('http') || url.startsWith('https')) {
+        window.open(url, '_blank');
```

Comment:
Why do we assume that absolute URLs need to open in a new tab? Just curiosity

Reply:
It seemed that the card details url varies, I follow their original implementation here and just update the route's state to avoid broken links:

https://github.com/elastic/kibana/pull/194028/files#diff-8b3cc54ced6b784857beabd4caf73fa9307b865cfa886e2d4c5652fbdcdb0374R169-R180